controllers/openstackmachine_controller.go

```diff
@@ -340,9 +364,33 @@ func (r *OpenStackMachineReconciler) reconcileNormal(ctx context.Context, scope
 		return ctrl.Result{}, err
 	}
 
-	instanceStatus, err := r.getOrCreate(scope.Logger(), cluster, openStackCluster, machine, openStackMachine, computeService, userData)
+	scope.Logger().Info("Adopting ports for OpenStackMachine", "name", openStackMachine.Name)
+	desiredPorts, err := networkingService.ConstructPorts(openStackCluster, openStackMachine.Spec.Ports, openStackMachine.Spec.Trunk)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("construct ports: %w", err)
+	}
+	if err := networkingService.AdoptPorts(openStackCluster, openStackMachine, desiredPorts); err != nil {
+		return ctrl.Result{}, fmt.Errorf("adopt ports: %w", err)
+	}
+
+	// Here I'm not sure how we want to find out what ports we need to create (the one that were not adopted)
+	// unless we want re-list the ports and compare with the ones we have in the status? (that would be a lot of API calls)
```

Comment:
ping @mdbooth

Reply:
Couple of things: firstly we should call AdoptPorts before either `reconcileNormal` or `reconcileDelete`. It's a separate phase before normal reconciliation.

AdoptPorts() should have ensured that status is accurate. Because we create ports in order, we know that if `len(desiredPorts) == 4` and `len(statusPorts) == 2` then we can skip over the first 2 and create ports 3 & 4.

Reply:
ok that clarify a bit, thanks.